aptos-move/move-examples/veiled_coin/sources/veiled_coin.move

```diff
@@ -0,0 +1,370 @@
+        let VeiledCoin<CoinType> { private_value: _ } = c;
+    }
+
+    //
+    // Public functions.
+    //
+
+    /// Returns `true` if `account_addr` is registered to receive veiled coins of `CoinType`.
+    public fun has_veiled_coin_store<CoinType>(account_addr: address): bool {
+        exists<VeiledCoinStore<CoinType>>(account_addr)
+    }
+
+    /// Returns the commitment to the value of `coin`.
+    public fun private_value<CoinType>(coin: &VeiledCoin<CoinType>): &Commitment {
+        &coin.private_value
+    }
+
+    /// Returns true if the balance at address `owner` equals `value`, which should be useful for auditability. Requires
+    /// the Pedersen commitment randomness as an auxiliary input.
+    public fun verify_opened_balance<CoinType>(owner: address, value: u64, randomness: &Scalar): bool acquires VeiledCoinStore {
+        // compute the expected committed balance
+        let value = new_scalar_from_u64(value);
+        let expected_comm = pedersen::new_commitment_for_bulletproof(&value, randomness);
+
+        // get the actual committed balance
+        let actual_comm = pedersen::new_commitment_from_compressed(&private_balance<CoinType>(owner));
+
+        pedersen::commitment_equals(&actual_comm, &expected_comm)
+    }
+
+    /// Initializes a veiled coin store for the specified account.
+    public fun register<CoinType>(account: &signer) {
+        let account_addr = signer::address_of(account);
+        assert!(
+            !has_veiled_coin_store<CoinType>(account_addr),
+            error::already_exists(EVEILED_COIN_STORE_ALREADY_PUBLISHED),
+        );
+
+        let coin_store = VeiledCoinStore<CoinType> {
+            private_balance: ristretto255::point_identity_compressed(),
+            deposit_events: account::new_event_handle<DepositEvent>(account),
+            withdraw_events: account::new_event_handle<WithdrawEvent>(account),
+        };
+        move_to(account, coin_store);
+    }
+
+    /// Mints a veiled coin from a normal coin, shelving the normal coin into the resource account's coin store.
+    ///
+    /// WARNING: Fundamentally, there is no way to hide the value of the coin being minted here.
+    public fun mint_from_coin<CoinType>(c: Coin<CoinType>): VeiledCoin<CoinType> acquires VeiledCoinMinter {
+        // If there is no CoinStore<CoinType> yet, create one.
+        let rsrc_acc_signer = get_resource_account_signer();
+        let rsrc_acc_addr = signer::address_of(&rsrc_acc_signer);
+        if(!coin::is_account_registered<CoinType>(rsrc_acc_addr)) {
+            coin::register<CoinType>(&rsrc_acc_signer);
+        };
+
+        // Move the traditional coin into the coin store, so we can mint a veiled coin.
+        // (There is no other way to drop a traditional coin, for safety reasons, so moving it into a coin store is
+        //  the only option.)
+        let value = new_scalar_from_u64(coin::value(&c));
+        coin::deposit(rsrc_acc_addr, c);
```

Comment:
This resource account owner can withdraw the coin and use them while the VeiledCoin exist. I think you might need a special coin escrow that guarantees no both veiled and non-veiled coin on the market.

Reply:
How would that happen? No one will know the SK of the resource account. The only way to touch that account is via the functions defined in this module, which are sufficiently restrictive. At least that's my understanding from @davidiw and @CapCap.

Reply:
@areshand, let me know if you still think there is a problem, before I mark this as resolve.d

Reply:
Seems safe to resolve... someone could always specify that the resource account has a known key per some of the helper functions in the library, but that can and should be audited, prior to anyone using this.